langchain/vectorstores/annoy.py

```diff
@@ -0,0 +1,414 @@
+        trees: int = 100,
+        n_jobs: int = -1,
+        **kwargs: Any,
+    ) -> Annoy:
+        """Construct Annoy wrapper from embeddings.
+
+        Args:
+            text_embeddings: List of tuples of (text, embedding)
+            embedding: Embedding function to use.
+            metadatas: List of metadata dictionaries to associate with documents.
+            metric: Metric to use for indexing. Defaults to "angular".
+            trees: Number of trees to use for indexing. Defaults to 100.
+            n_jobs: Number of jobs to use for indexing. Defaults to -1
+
+        This is a user friendly interface that:
+            1. Creates an in memory docstore with provided embeddings
+            2. Initializes the Annoy database
+
+        This is intended to be a quick way to get started.
+
+        Example:
+            .. code-block:: python
+
+                from langchain import Annoy
+                from langchain.embeddings import OpenAIEmbeddings
+                embeddings = OpenAIEmbeddings()
+                db = Annoy.from_texts(texts, embeddings)
+        """
+        texts = [t[0] for t in text_embeddings]
+        embeddings = [t[1] for t in text_embeddings]
+
+        return cls.__from(
+            texts, embeddings, embedding, metadatas, metric, trees, n_jobs, **kwargs
+        )
+
+    def save_local(self, folder_path: str, prefault: bool = False) -> None:
+        """Save Annoy index, docstore, and index_to_docstore_id to disk.
+
+        Args:
+            folder_path: folder path to save index, docstore,
+                and index_to_docstore_id to.
+            prefault: Whether to pre-load the index into memory.
+        """
+        path = Path(folder_path)
+        os.makedirs(path, exist_ok=True)
+        # save index
+        self.index.save(str(path / "index.annoy"), prefault=prefault)
+        # save docstore and index_to_docstore_id
+        with open(path / "index.pkl", "wb") as file:
+            pickle.dump((self.docstore, self.index_to_docstore_id), file)
+
+    @classmethod
+    def load_local(
+        cls,
+        folder_path: str,
+        embeddings: Embeddings,
+        f: int = 0,
+        metric: str = DEFAULT_METRIC,
+    ) -> Annoy:
+        """Load Annoy index, docstore, and index_to_docstore_id to disk.
+
+        Args:
+            folder_path: folder path to load index, docstore,
+                and index_to_docstore_id from.
+            embeddings: Embeddings to use when generating queries.
+            f: Number of dimensions in the embeddings.
+            metric: Metric to use for the index.
+        """
+        assert metric in INDEX_METRICS, f"Invalid metric {metric}"
+        path = Path(folder_path)
+        # load index separately since it is not picklable
+        annoy = dependable_annoy_import()
+        # derive embedding dimensionality from random doc if not provided
+        if not f:
+            f = len(embeddings.embed_documents(["dimension"])[0])
```

Comment:
Rather than re-compute the embedding dimension, can we save a `config.json` in the index directory with the metric and embedding dimension? This vector store is read-only anyway

Reply:
yes, indeed: storing the config is the obvious way to go - with that,  `load_local` relies only on the path and the embedding method. I used configparser and added it to the pickle